### ISSUE TITLE
A way to get deployment feedback

### ISSUE THREAD

**Issue by zadigus** [devspace-cloud/devspace#954]
**Is your feature request related to a problem?**  
In my CI / CD pipeline, I would like an easy way to get feedback on how my deployment is doing. Indeed, in such a pipeline, I want e.g. to build, unit test, deploy to staging, test my staging environment, deploy to production. To do so, I need a way to know when I can start my staging tests, i.e. I need the guarantee that the relevant services are up and running so that my tests can be run on the system I expect. 

**Which solution do you suggest?**  
I'm not an expert of k8s, but I guess something along the lines
```
kubectl rollout status
```
or 
```
helm --wait
```
could do the trick.

**Which alternative solutions exist?**  
Currently, I'm not entirely working with devspace, I'm evaluating it. My solution to this problem is to have a [skaffold](skaffold.dev) configuration and run
```
skaffold run --status-check
```

**Additional context**  

/kind feature

**Comment by FabianKramm** [comment]
@zadigus thanks for opening this issue! Have you tried `devspace analyze` and `devspace list deployments`? Devspace analyze will check for problems in the namespace and also verify the pods are running and not in a failure state.

**Comment by zadigus** [comment]
No, I haven't. I'll try that out and come back to you. Thanks for the advice!

**Comment by zadigus** [comment]
So, first attempt. I am deploying a `bitnami/postgresql-ha`. After the call
```
devspace -n devspace deploy
```
I run
```
$ devspace analyze -n devspace
[info]   Using kube context 'docker-desktop'
[info]   Using namespace 'devspace'
```
with the result
```
  ================================================================================
                           Events (3 potential issue(s))                          
  ================================================================================
  Warning (34m30s ago) - Pod minio-initialization-expose-buckets-7blqq: 
  1x Error: cannot find volume "default-token-l4k87" to mount into container "wait-for-minio"

  Warning (15s ago) - Pod postgres-pgpool-664c58f7c8-r2dg8: 
  5x Readiness probe failed: psql: FATAL:  failed to create a backend 0 connection
DETAIL:  not executing failover because failover_on_backend_error is off


  Warning (2s ago) - Pod postgres-postgresql-1: 
  1x Readiness probe failed: psql: could not connect to server: Connection refused
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 5432?
```
However, this is non-blocking. I need to have a command that really waits for the deployment to be done on my ci / cd. Otherwise, how does my pipeline know that it can start with the tests on the deployed services? I could write a loop running that `devspace analyze` command, but it'd be pretty annoying. In that, I think the `skaffold run --status-check` is much more powerful. As long as the deployments are not done or a timeout has been reached, the command does not return. Do you have currently something going into that direction?

**Comment by FabianKramm** [comment]
@zadigus I see, so essentially you want to deploy everything in the beginning (thats why you cannot use `deployments.*.helm.wait` and then wait for a certain timeout till all the deployments become eventually ready, is that correct?

**Comment by zadigus** [comment]
@FabianKramm yes, it is. After some discussions with you or your colleague on the support chat, I came up with the following (yet not perfect) script that I can call in the devspace hooks, after all deployments:
```
#! /bin/bash

if [ "$#" -lt "1" ] ; then
  echo "Usage: $0 namespace [timeout_iterations]"
  exit 1
fi

NAMESPACE=$1
TIMEOUT=${2:-300}

deployment_names=$(kubectl get deployments -o jsonpath='{.items[*].metadata.name}' -n $NAMESPACE)
nb_deployments=$(echo ${deployment_names} | wc -w)

echo "Waiting for deployments to be done"
for iteration in `seq 1 $TIMEOUT`; do
  nb_success=0
  for name in $deployment_names ; do
    status=$(kubectl rollout status deployment $name --namespace $NAMESPACE --watch=false | grep "success" | wc -l)
    if [ "$status" -eq "1" ] ; then
      echo "Deployment $name rolled out."
    fi
    nb_success=$(($nb_success + $status))
  done
  if [ "$nb_success" -eq "$nb_deployments" ] ; then
    echo "Deployments rolled out successfully"
    exit 0
  else
    echo "Iteration $iteration"
    sleep 1
  fi
done
echo "Deployments timed out" && exit 1
```
That script's outputs are definitely less user-friendly than the devspace outputs. It should also, in my personal case, wait until all jobs are done, which is a pretty easy thing to add.

**Comment by zadigus** [comment]
I just have one issue with the above script. I've stored it under `scripts/wait_for_deployments.sh`. I am working with git bash in my dev environment. Defining the hook like this doesn't work:
```
hooks:
- command: ./scripts/wait_for_deployments.sh 
  args:
    - devspace
  when:
    after:
      deployments: all
```
I get the following error:
```
[fatal]  Error executing hook: fork/exec ./scripts/wait_for_deployments.sh: %1 is not a valid Win32 application.
```
How do I make it work in git bash and in any linux terminal?

**Comment by FabianKramm** [comment]
@zadigus thanks for the additional information. After some consideration, I think we will add a new `--wait` flag to `devspace deploy` that is basically doing what you want in that it waits for a certain timeout (specified via `--timeout`) for all the deployments / replicasets in a given namespace and we also add a `devspace analyze --patient` flag that ignores any error until the given timeout, which should solve your problem.  

Regarding the hook, the best practice is to create a separate script for windows and linux. In your case your windows script would just call the linux script via git bash and your config would look like this:

```yaml
hooks:
- command: ./scripts/wait_for_deployments
  args:
    - devspace
  when:
    after:
      deployments: all
```

Then you would have a script called `wait_for_deployments` (the linux one) and a script called `wait_for_deployments.cmd` (the windows one) that just calls the other script via git bash.

**Comment by FabianKramm** [comment]
Implemented in #973
